packages/SwingSet/src/vats/comms/gc-comms.js

```diff
@@ -0,0 +1,324 @@
+import { assert } from '@agoric/assert';
+import { parseVatSlot } from '../../parseVatSlots.js';
+import { parseRemoteSlot } from './parseRemoteSlot.js';
+
+// A note on verb polarity:
+//
+// On any vat-kernel or comms-comms boundary, the "exporter" is the side who
+// first sent the object reference, and thus "owns" the object: any messages
+// sent to the object are sent to the exporting side. Think of this as
+// "upstream". The side which first *received* the reference is the importer,
+// think of them as downstream.
+//
+// On the vat-kernel boundary, where there is a clear symmetry break, our
+// rule is "vats are egocentric". Vats think of themselves as the center of
+// the universe, so naturally vat exports are more interesting than anything
+// else, so the vat assigns positive identifiers to exported object
+// references (i.e. `o+0` is the exported root object, `o+1` is the first
+// non-root export). When the vat *imports* something from the kernel, it
+// receives a negative identifier (i.e. the kernel might send `o-2` in the
+// argument of a delivery).
+//
+// Likewise, the GC operations are described from the vat's point of view.
+// When the vat no longer needs an imported object, it calls
+// `syscall.dropImport` to tell the kernel "I'm dropping this thing I
+// imported from you". `syscall.retireImport` is how the vat tells the kernel
+// it can no longer recognize the object and the c-list entry should be
+// deleted. Both messages flow "upstream" from the importing vat to the
+// kernel. Then, if all importers have dropped a kernel object, and nothing
+// else in the kernel can reach it, the kernel will send
+// `dispatch.dropExport` and/or `dispatch.retireExport` to inform the
```

Comment:
It just now occurs to me to ask if `retireExport` subsumes `dropExport`.  That is, does a `retireExport` need to be preceded by a `dropExport` or is that implied automatically?

Reply:
The drop must appear before the retire, but the retire does not imply the drop.

```
state=REACHABLE     <------------------\
 |                                     |
 | dispatch.dropExport                 |
 v                                     |
state=RECOGNIZABLE   ---(new import)->-/
 |                                     |
 | dispatch.retireExport               |
 v                                     |
state=NONE           ---(new import)->-/
```